File: backend/app/services/maintenance.py
import asyncio
from contextlib import suppress
from datetime import timedelta

from sqlalchemy import delete, func, select
from sqlalchemy.orm import selectinload

from app.db.models import (
    Channel,
    DiscoveryRun,
    HealthProbeLog,
    RequestAttempt,
    RequestLog,
    utcnow,
)
from app.services.circuit_breaker import as_utc
from app.services.discovery import discover_channel_models
from app.services.settings import get_runtime_settings


async def reconcile_completed_stream_cancellations(session_factory) -> int:
    """Repair legacy cancellations that already captured a completed stream's usage."""
    async with session_factory() as session:
        rows = (
            (
                await session.execute(
                    select(RequestLog)
                    .options(selectinload(RequestLog.attempts))
                    .where(
                        RequestLog.outcome == "cancelled",
                        RequestLog.final_status_code == 200,
                        RequestLog.response_bytes > 0,
                    )
                )
            )
            .scalars()
            .all()
        )
        repaired = 0
        for row in rows:
            if not row.attempts:
                continue
            final_attempt = max(row.attempts, key=lambda attempt: attempt.attempt_no)
            if not (
                final_attempt.outcome == "cancelled"
                and final_attempt.status_code == 200
                and final_attempt.response_started
                and final_attempt.raw_usage_json is not None
                and final_attempt.output_tokens is not None
            ):
                continue
            row.outcome = "success"
            final_attempt.outcome = "success"
            repaired += 1
        if repaired:
            await session.commit()
        return repaired


class MaintenanceSupervisor:
    def __init__(self, app, interval_seconds: int = 60) -> None:
        self.app = app
        self.interval_seconds = interval_seconds
        self._task: asyncio.Task | None = None
        self._discovering: set[str] = set()
        self._last_cleanup = None

    def start(self) -> None:
        self._task = asyncio.create_task(self._run(), name="maintenance-supervisor")

    async def stop(self) -> None:
        if self._task:
            self._task.cancel()
            with suppress(asyncio.CancelledError):
                await self._task

    async def _run(self) -> None:
        while True:
            try:
                await self._schedule_discovery()
                await self._finalize_stale_pending_requests()
                await self._cleanup_logs()
            except Exception:
                pass
            await asyncio.sleep(self.interval_seconds)

    async def _schedule_discovery(self) -> None:
        now = utcnow()
        async with self.app.state.db.sessions() as session:
            runtime = await get_runtime_settings(session)
            cutoff = now - timedelta(hours=int(runtime["model_discovery_interval_hours"]))
            channels = (
                (await session.execute(select(Channel.id).where(Channel.manual_enabled.is_(True))))
                .scalars()
                .all()
            )
            due: list[str] = []
            for channel_id in channels:
                last_started = await session.scalar(
                    select(func.max(DiscoveryRun.started_at)).where(
                        DiscoveryRun.channel_id == channel_id
                    )
                )
                if last_started is None or as_utc(last_started) <= cutoff:
                    due.append(channel_id)
            runs: list[tuple[str, str]] = []
            for channel_id in due:
                if channel_id in self._discovering:
                    continue
                run = DiscoveryRun(channel_id=channel_id, trigger="scheduled")
                session.add(run)
                await session.flush()
                runs.append((channel_id, run.id))
                self._discovering.add(channel_id)
            await session.commit()
        for channel_id, run_id in runs:
            asyncio.create_task(self._run_discovery(channel_id, run_id))

    async def _run_discovery(self, channel_id: str, run_id: str) -> None:
        try:
            await discover_channel_models(
                self.app.state.db.sessions,
                self.app.state.http,
                self.app.state.secrets,
                channel_id,
                run_id,
            )
        finally:
            self._discovering.discard(channel_id)

    async def _finalize_stale_pending_requests(self) -> None:
        now = utcnow()
        async with self.app.state.db.sessions() as session:
            runtime = await get_runtime_settings(session)
            stale_seconds = max(
                int(runtime["stream_idle_timeout_seconds"]) * 2,
                int(runtime["first_byte_timeout_seconds"]) * 2,
                600,
            )
            cutoff = now - timedelta(seconds=stale_seconds)
            rows = (
                (
                    await session.execute(
                        select(RequestLog).where(
                            RequestLog.outcome == "pending",
                            RequestLog.started_at < cutoff,
                        )
                    )
                )
                .scalars()
                .all()
            )
            for row in rows:
                started_at = as_utc(row.started_at)
                row.finished_at = now
                row.total_duration_ms = (
                    round((now - started_at).total_seconds() * 1000) if started_at else None
                )
                row.outcome = "cancelled"
                row.response_bytes = row.response_bytes or 0
            await session.commit()

    async def _cleanup_logs(self) -> None:
        now = utcnow()
        if self._last_cleanup and now - self._last_cleanup < timedelta(hours=1):
            return
        async with self.app.state.db.sessions() as session:
            runtime = await get_runtime_settings(session)
            cutoff = now - timedelta(days=int(runtime["log_retention_days"]))
            expired_requests = select(RequestLog.id).where(RequestLog.started_at < cutoff)
            await session.execute(
                delete(RequestAttempt).where(RequestAttempt.request_id.in_(expired_requests))
            )
            await session.execute(delete(RequestLog).where(RequestLog.started_at < cutoff))
            await session.execute(delete(HealthProbeLog).where(HealthProbeLog.started_at < cutoff))
            await session.execute(delete(DiscoveryRun).where(DiscoveryRun.started_at < cutoff))
            await session.commit()
        self._last_cleanup = now

File: desktop/src-tauri/src/maintenance.rs
//! Background maintenance — port of backend/app/services/maintenance.py (C4).
//!
//! Runs a 60-second loop that:
//! 1. schedules model discovery for channels whose last run is older than
//!    `model_discovery_interval_hours`,
//! 2. finalizes request logs stuck in `pending` (cancelled, with a computed
//!    duration),
//! 3. deletes logs/probes/runs older than `log_retention_days` (hourly).
//!
//! `reconcile_completed_stream_cancellations` repairs legacy rows where a
//! fully completed stream was recorded as cancelled; it runs once at startup.

use std::collections::HashMap;
use std::sync::Arc;

use sqlx::Row;
use tokio::sync::Mutex;
use tokio::time::{Duration, Instant};

use crate::server::AppState;
use crate::settings;

/// Final attempt bookkeeping read back from `request_attempts` for repair:
/// (outcome, status_code, response_started, has_raw_usage, output_tokens).
type FinalAttempt = (String, Option<i64>, Option<i64>, bool, Option<i64>);

/// Repair legacy cancellations that already captured a completed stream's
/// usage (Python `reconcile_completed_stream_cancellations`).
pub async fn reconcile_completed_stream_cancellations(state: &AppState) -> anyhow::Result<i64> {
    let rows: Vec<(String, String)> = sqlx::query(
        "SELECT rl.id, rl.response_bytes FROM request_logs rl \
         WHERE rl.outcome = 'cancelled' AND rl.final_status_code = 200 AND rl.response_bytes > 0",
    )
    .fetch_all(state.db.pool())
    .await?
    .into_iter()
    .map(|row| {
        (
            row.try_get("id").unwrap_or_default(),
            row.try_get("response_bytes").unwrap_or_default(),
        )
    })
    .collect();
    let mut repaired = 0i64;
    for (request_id, _) in rows {
        let final_attempt: Option<FinalAttempt> = sqlx::query(
            "SELECT outcome, status_code, response_started, raw_usage_json, output_tokens \
                 FROM request_attempts WHERE request_id = ? ORDER BY attempt_no DESC LIMIT 1",
        )
        .bind(&request_id)
        .fetch_optional(state.db.pool())
        .await?
        .map(|row| -> anyhow::Result<FinalAttempt> {
            Ok((
                row.try_get::<String, _>("outcome")?,
                row.try_get::<Option<i64>, _>("status_code")?,
                row.try_get::<Option<bool>, _>("response_started")?
                    .map(|v| v as i64),
                row.try_get::<Option<String>, _>("raw_usage_json")?
                    .is_some(),
                row.try_get::<Option<i64>, _>("output_tokens")?,
            ))
        })
        .transpose()?;
        let Some((outcome, status_code, response_started, raw_usage, output_tokens)) =
            final_attempt
        else {
            continue;
        };
        if outcome == "cancelled"
            && status_code == Some(200)
            && response_started == Some(1)
            && raw_usage
            && output_tokens.is_some()
        {
            sqlx::query("UPDATE request_logs SET outcome='success' WHERE id=?")
                .bind(&request_id)
                .execute(state.db.pool())
                .await?;
            sqlx::query(
                "UPDATE request_attempts SET outcome='success' \
                 WHERE request_id=? AND attempt_no=(SELECT MAX(attempt_no) FROM request_attempts WHERE request_id=?)",
            )
            .bind(&request_id)
            .bind(&request_id)
            .execute(state.db.pool())
            .await?;
            repaired += 1;
        }
    }
    Ok(repaired)
}

/// Background supervisor (Python `MaintenanceSupervisor`, 60s cadence).
pub fn spawn_supervisor(state: AppState) {
    tokio::spawn(async move {
        let _ = reconcile_completed_stream_cancellations(&state).await;
        let discovering: Arc<Mutex<HashMap<String, String>>> = Arc::new(Mutex::new(HashMap::new()));
        let mut last_cleanup: Option<Instant> = None;
        let mut interval = tokio::time::interval(Duration::from_secs(60));
        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
        loop {
            interval.tick().await;
            if let Err(error) = schedule_discovery(&state, &discovering).await {
                tracing::warn!(%error, "scheduled discovery failed");
            }
            if let Err(error) = finalize_stale_pending_requests(&state).await {
                tracing::warn!(%error, "stale request finalization failed");
            }
            let due = last_cleanup.is_none_or(|last| last.elapsed() >= Duration::from_secs(3600));
            if due {
                if let Err(error) = cleanup_logs(&state).await {
                    tracing::warn!(%error, "log cleanup failed");
                }
                last_cleanup = Some(Instant::now());
            }
        }
    });
}

/// Create `scheduled` discovery runs for enabled channels whose last run is
/// older than `model_discovery_interval_hours` and spawn the background work.
async fn schedule_discovery(
    state: &AppState,
    discovering: &Arc<Mutex<HashMap<String, String>>>,
) -> anyhow::Result<()> {
    let runtime = settings::runtime_settings(state).await?;
    let interval_hours = runtime.model_discovery_interval_hours.max(1);
    let cutoff = (chrono::Utc::now() - chrono::Duration::hours(interval_hours)).to_rfc3339();
    let channels: Vec<String> =
        sqlx::query_scalar("SELECT id FROM channels WHERE manual_enabled = 1 ORDER BY id")
            .fetch_all(state.db.pool())
            .await?;
    let mut due: Vec<String> = Vec::new();
    for channel_id in channels {
        let last_started: Option<String> =
            sqlx::query_scalar("SELECT MAX(started_at) FROM discovery_runs WHERE channel_id = ?")
                .bind(&channel_id)
                .fetch_optional(state.db.pool())
                .await?;
        let needs_run = match last_started {
            None => true,
            Some(last) => last <= cutoff,
        };
        if needs_run {
            due.push(channel_id);
        }
    }
    let mut discovering = discovering.lock().await;
    // Drop channels whose queued run already finished.
    let mut finished = Vec::new();
    for (channel_id, run_id) in discovering.iter() {
        let done: Option<Option<String>> =
            sqlx::query_scalar("SELECT finished_at FROM discovery_runs WHERE id = ?")
                .bind(run_id)
                .fetch_optional(state.db.pool())
                .await?;
        if done.flatten().is_some() {
            finished.push(channel_id.clone());
        }
    }
    for channel_id in finished {
        discovering.remove(&channel_id);
    }
    for channel_id in due {
        if discovering.contains_key(&channel_id) {
            continue;
        }
        let run_id =
            match crate::discovery::queue_scheduled(state.clone(), channel_id.clone()).await {
                Ok(run_id) => run_id,
                Err(error) => {
                    tracing::warn!(%error, "failed to queue scheduled discovery");
                    continue;
                }
            };
        discovering.insert(channel_id, run_id);
    }
    Ok(())
}

/// Mark request logs stuck in `pending` as cancelled (Python
/// `_finalize_stale_pending_requests`).
async fn finalize_stale_pending_requests(state: &AppState) -> anyhow::Result<()> {
    let runtime = settings::runtime_settings(state).await?;
    let stale_seconds = runtime
        .stream_idle_timeout_seconds
        .max(runtime.first_byte_timeout_seconds)
        .max(300)
        * 2;
    let cutoff = (chrono::Utc::now() - chrono::Duration::seconds(stale_seconds)).to_rfc3339();
    let rows: Vec<(String, Option<String>)> = sqlx::query(
        "SELECT id, started_at FROM request_logs \
         WHERE outcome = 'pending' AND started_at < ?",
    )
    .bind(&cutoff)
    .fetch_all(state.db.pool())
    .await?
    .into_iter()
    .map(|row| {
        Ok((
            row.try_get::<String, _>("id")?,
            row.try_get::<Option<String>, _>("started_at")?,
        ))
    })
    .collect::<anyhow::Result<Vec<_>>>()?;
    let now = chrono::Utc::now();
    for (request_id, started_at) in rows {
        let duration_ms = started_at
            .and_then(|started| chrono::DateTime::parse_from_rfc3339(&started).ok())
            .map(|started| {
                (now - started.with_timezone(&chrono::Utc))
                    .num_milliseconds()
                    .max(0)
            });
        sqlx::query(
            "UPDATE request_logs SET finished_at=?, total_duration_ms=?, outcome='cancelled', \
             response_bytes=COALESCE(response_bytes,0) WHERE id=?",
        )
        .bind(now.to_rfc3339())
        .bind(duration_ms)
        .bind(&request_id)
        .execute(state.db.pool())
        .await?;
    }
    Ok(())
}

/// Delete requests, attempts, probes and discovery runs older than
/// `log_retention_days` (Python `_cleanup_logs`, at most once per hour).
async fn cleanup_logs(state: &AppState) -> anyhow::Result<()> {
    let runtime = settings::runtime_settings(state).await?;
    let cutoff = (chrono::Utc::now() - chrono::Duration::days(runtime.log_retention_days.max(1)))
        .to_rfc3339();
    let mut tx = state.db.pool().begin().await?;
    sqlx::query(
        "DELETE FROM request_attempts WHERE request_id IN (SELECT id FROM request_logs WHERE started_at < ?)",
    )
    .bind(&cutoff)
    .execute(&mut *tx)
    .await?;
    sqlx::query("DELETE FROM request_logs WHERE started_at < ?")
        .bind(&cutoff)
        .execute(&mut *tx)
        .await?;
    sqlx::query("DELETE FROM health_probe_logs WHERE started_at < ?")
        .bind(&cutoff)
        .execute(&mut *tx)
        .await?;
    sqlx::query("DELETE FROM discovery_runs WHERE started_at < ?")
        .bind(&cutoff)
        .execute(&mut *tx)
        .await?;
    tx.commit().await?;
    Ok(())
}
